### ISSUE TITLE
Assignment to Final attribute incorrectly flagged in conditional assignment within __init__

### ISSUE THREAD

**Issue by nernst** [python/mypy#20443]
**Bug Report**

Documentation on `typing.Final` says (emphasis mine):

 Final names cannot be *reassigned* in any scope. Final names declared in class scopes cannot be overridden in subclasses.

The problem comes in a class's init, if the final member is conditionally assigned to different values (e.g. assignments in an if/else expression, if the assignment to the member occurs in both blocks). In the if/else case, mypy flags the assignment in the else block.

**To Reproduce**

```python
from typing import Final


class Foo:
    attr: Final[str]

    def __init__(self, flag: bool) -> None:
        if flag:
            self.attr = "bar"
        else:
            self.attr = "baz"  # mypy flags with:  error: Cannot assign to final attribute "attr"  [misc]
```

**Expected Behavior**

I would expect mypy to not flag this code at all. For either possible value of `flag`, there is only a single assignment, so no _reassignment_ occurs.

**Actual Behavior**

Mypy flags line 11 with: `error: Cannot assign to final attribute "attr"  [misc]`

**Your Environment**

- Mypy version used: mypy 1.19.1 (compiled: yes)
- Mypy command-line flags: none; just the path to the repro test script
- Mypy configuration options from `mypy.ini` (and other config files):
```toml
[tool.mypy]
python_version = "3.13"

# set mypy_path and explicit_package_bases, otherwise mypy gets confused about package names
# due to the use of namespace packages.
mypy_path = "src"
explicit_package_bases = true

pretty = true
disallow_any_unimported = true
disallow_any_expr = false
disallow_any_explicit = false
disallow_any_generics = true
disallow_subclassing_any = true
disallow_untyped_calls = true
disallow_untyped_defs = true
disallow_incomplete_defs = true
disallow_untyped_decorators = true
implicit_optional = true
strict_optional = true
warn_redundant_casts = true
warn_unused_ignores = true
warn_no_return = true
warn_return_any = true
warn_unreachable = true
warn_unused_configs = true
exclude = ["dist/.*/.*\\.py$"]
```
- Python version used: 3.13.7

**Comment by sterliakov** [comment]
I thing this adheres to the spirit of Final, though the spec does not mention such a case, and the language is rather clear:

> There must be exactly one assignment to a final name.

This should probably be allowed, if/else assignment is semantically equivalent to assignment of a ternary, and only one assignment actually happens at runtime. Might not be trivial to implement, though: now we just reject any assignment to something with a `Final` qualifier, so allowing this would require some care.
